// @platejs/basic-nodes/dist/index.js
import { _ as BaseH6Plugin, a as BaseSuperscriptPlugin, c as BaseItalicPlugin, d as BaseHorizontalRulePlugin, f as BaseH1Plugin, g as BaseH5Plugin, h as BaseH4Plugin, i as BaseUnderlinePlugin, l as BaseCodePlugin, m as BaseH3Plugin, n as BaseHighlightPlugin, o as BaseSubscriptPlugin, p as BaseH2Plugin, r as BaseBasicMarksPlugin, s as BaseStrikethroughPlugin, t as BaseKbdPlugin, u as BaseBoldPlugin, v as BaseHeadingPlugin, y as BaseBlockquotePlugin } from "./BaseKbdPlugin-myd8t1-T.js";
import { createSlatePlugin } from "platejs";

//#region src/lib/BaseBasicBlocksPlugin.ts
const BaseBasicBlocksPlugin = createSlatePlugin({ plugins: [
	BaseBlockquotePlugin,
	BaseHeadingPlugin,
	BaseHorizontalRulePlugin
] });

//#endregion
export { BaseBasicBlocksPlugin, BaseBasicMarksPlugin, BaseBlockquotePlugin, BaseBoldPlugin, BaseCodePlugin, BaseH1Plugin, BaseH2Plugin, BaseH3Plugin, BaseH4Plugin, BaseH5Plugin, BaseH6Plugin, BaseHeadingPlugin, BaseHighlightPlugin, BaseHorizontalRulePlugin, BaseItalicPlugin, BaseKbdPlugin, BaseStrikethroughPlugin, BaseSubscriptPlugin, BaseSuperscriptPlugin, BaseUnderlinePlugin };
